### PR TITLE
[LV, VPlan] Check if plan is compatible to EVL transform

### PR DESCRIPTION
**Pull request by arcbbb**
The transform updates all users of inductions to work based on EVL, instead
of the VF directly. At the moment, widened inductions cannot be updated, so
bail out if the plan contains any.
This patch introduces a check before applying EVL transform. If any recipes in loop rely on RuntimeVF, the plan is discarded.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -8539,6 +8539,29 @@ VPRecipeBuilder::tryToCreateWidenRecipe(Instruction *Instr,
...
+        return true;
+    for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+             vp_depth_first_deep(Plan.getVectorLoopRegion())))
+      for (VPRecipeBase &Recipe : *VPBB)
+        if (auto *VecPtrR = dyn_cast<VPVectorPointerRecipe>(&Recipe))
+          if (VecPtrR->isReverse())
+            return true;
+    return false;
+  };
+  if (HasAnyRuntimeVFUserInLoop(Plan))
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8560]: Simpler to not use the lambda but directly runtime from the loops in it?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -8539,6 +8539,29 @@ VPRecipeBuilder::tryToCreateWidenRecipe(Instruction *Instr,
   return tryToWiden(Instr, Operands, VPBB);
 }
 
+// EVL transform doesn't support backends where EVL diffs from RuntimeVF
+// in the second-to-last iteration.
+// Return false if the vector region has recipes relying on
+// RuntimeVF.
+static bool isCompatibleToEVLTransform(VPlan &Plan) {
+  auto HasAnyRuntimeVFUserInLoop = [](VPlan &Plan) -> bool {
+    for (auto &Phi : Plan.getVectorLoopRegion()->getEntryBasicBlock()->phis())
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8548]: nit: use any_of? also, what about users of the canonical induction?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -8539,6 +8539,29 @@ VPRecipeBuilder::tryToCreateWidenRecipe(Instruction *Instr,
...
+      if (isa<VPWidenIntOrFpInductionRecipe>(&Phi) ||
+          isa<VPWidenPointerInductionRecipe>(&Phi))
+        return true;
+    for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+             vp_depth_first_deep(Plan.getVectorLoopRegion())))
+      for (VPRecipeBase &Recipe : *VPBB)
+        if (auto *VecPtrR = dyn_cast<VPVectorPointerRecipe>(&Recipe))
+          if (VecPtrR->isReverse())
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8556]: Why is this needed? If it is not supported we should not even attempt vectorizing with EVL? cc @alexey-bataev

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 4]
```diff
@@ -8553,8 +8576,12 @@ void LoopVectorizationPlanner::buildVPlansWithVPRecipes(ElementCount MinVF,
             *Plan, CM.getMinimalBitwidths(), PSE.getSE()->getContext());
       VPlanTransforms::optimize(*Plan, *PSE.getSE());
       // TODO: try to put it close to addActiveLaneMask().
-      if (CM.foldTailWithEVL())
+      if (CM.foldTailWithEVL()) {
+        // Don't generate plan if the plan is not EVL-compatible
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8580]: Nit: the plan has already been generated, maybe 'discard` would be more accurate.

File: llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll [hunk 5]
```diff
@@ -0,0 +1,103 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -passes=loop-vectorize -force-tail-folding-style=data-with-evl \
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll:2]: please add the tests in a separate commit/PR, so the diff shows the changed by this patch only

Comment (arcbbb) [reply llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll:2]: Created a separate PR at https://github.com/llvm/llvm-project/pull/92203

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 4 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Fixed. Thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 3 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Removed. I didn't notice that foldTailWithEVL would check for CM_Widen_Reverse. Thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 2 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Fixed. Since EVL transform would replace all uses of CanonicalIV with EVLPhi, I think it is ok for now..

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 6]
```diff
@@ -8539,6 +8539,20 @@ VPRecipeBuilder::tryToCreateWidenRecipe(Instruction *Instr,
...
+// in the second-to-last iteration.
+// Return false if the vector region has recipes relying on
+// RuntimeVF.
+static bool isCompatibleToEVLTransform(VPlan &Plan) {
+  if (any_of(Plan.getVectorLoopRegion()->getEntryBasicBlock()->phis(),
+             [](VPRecipeBase &Phi) {
+               return (isa<VPWidenIntOrFpInductionRecipe>(&Phi) ||
+                       isa<VPWidenPointerInductionRecipe>(&Phi));
+             }))
+    return false;
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8552]: nit: can replace `if()` with direct `return` (possibly needing to negate/or use `all_of`. With it being that compact, it would probably be better to directly fold this into the transform that introduces EVL

Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8552]: OK, let's try this way.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 7]
```diff
@@ -1306,7 +1306,16 @@ void VPlanTransforms::addActiveLaneMask(
 /// %NextEVLIV = add IVSize (cast i32 %VPEVVL to IVSize), %EVLPhi
 /// ...
 ///
-void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+bool VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+  // EVL transform doesn't support backends where EVL diffs from RuntimeVF
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1310]: Could you re-phrase the comment. It's not clear what backend support here means. Perhaps something like

```
The transform updates all users of inductions to work based on EVL, instead of the VF directly. At the moment, widened inductions cannot be updated, so bail out if the plan contains any.
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 8]
```diff
@@ -1306,7 +1306,16 @@ void VPlanTransforms::addActiveLaneMask(
...
+  // in the second-to-last iteration.
+  // Return false if any recipes rely on RuntimeVF.
+  if (any_of(Plan.getVectorLoopRegion()->getEntryBasicBlock()->phis(),
+             [](VPRecipeBase &Phi) {
+               return (isa<VPWidenIntOrFpInductionRecipe>(&Phi) ||
+                       isa<VPWidenPointerInductionRecipe>(&Phi));
+             }))
+    return false;
   VPBasicBlock *Header = Plan.getVectorLoopRegion()->getEntryBasicBlock();
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1319]: Move up and use 

```
if (any_of(Header->phis())...
```

File: llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+; RUN: -prefer-predicate-over-epilogue=predicate-dont-vectorize \
+; RUN: -force-target-instruction-cost=1 -mtriple=riscv64 -mattr=+v -S < %s | FileCheck %s
+
+; Check loops having VPWidenIntOrFpInductionRecipe
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll:6]: nit: would be good to have an explanation, e.g. `Make sure we do not vectorize a loop with a widened induction.`

Same for the others

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 8 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1310]: Fixed. Thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 7 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Fixed. Thanks!

File: llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll (hunk 9 above)
Comment (arcbbb) [reply llvm/test/Transforms/LoopVectorize/RISCV/evl-compatible-loops.ll]: Fixed and also updated #92203

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.h [hunk 10]
```diff
@@ -104,7 +104,8 @@ struct VPlanTransforms {
   /// VPCanonicalIVPHIRecipe with a VPEVLBasedIVPHIRecipe.
   /// VPCanonicalIVPHIRecipe is only used to control the loop after
   /// this transformation.
-  static void addExplicitVectorLength(VPlan &Plan);
+  /// \returns true if the transformation succeeds, or false if it doesn't.
+  static bool addExplicitVectorLength(VPlan &Plan);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.h:108]: nit: `tryAddExplicitVectorLength` would be more accurate now?

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

> Vector loops generated from the EVL transform may experience issues on architectures where EVL differs from RuntimeVF at the second-to-last iteration.
> 

For the part above, I'd use the same wording as the commit in addExplicitVectorLength

> This patch introduces a check before applying EVL transform. If any recipes in loop rely on RuntimeVF, the vectorizer stop to generate the plan.

the vectorizer stop to generate the plan -> the plan is discarded?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.h (hunk 10 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.h]: Fixed. Thanks!

**Comment by wangpc-pp** [comment]
I found some loops can't be vectorized and finally I dug to this PR. :-)
I simply commented out these lines in this PR and the loop can be vectorized correctly. What's the status now? Are there any things I missed?

**Comment by zixuan-wu** [comment]
> I found some loops can't be vectorized and finally I dug to this PR. :-) I simply commented out these lines in this PR and the loop can be vectorized correctly. What's the status now? Are there any things I missed?

Do you find any other cases with functional/performance issue if it bails out to DataWithoutLaneMask style mask? Is it better to bail out to TailFoldingStyle::Data style as what getPreferredTailFoldingStyle prefers when V extension enables. If it bails out to TailFoldingStyle::Data style, then useActiveLaneMask should do some work at tryToBuildVPlanWithVPRecipes.

**Comment by wangpc-pp** [comment]
> > I found some loops can't be vectorized and finally I dug to this PR. :-) I simply commented out these lines in this PR and the loop can be vectorized correctly. What's the status now? Are there any things I missed?
> 
> Do you find any other cases with functional/performance issue if it bails out to DataWithoutLaneMask style mask instead of discarding the VPlan? Is it better to bail out to TailFoldingStyle::Data style as what getPreferredTailFoldingStyle prefers when V extension enables. If it bails out to TailFoldingStyle::Data style, then useActiveLaneMask should do some work at tryToBuildVPlanWithVPRecipes.

The problem I met is because there are some `VPWidenPointerInductionRecipe`s and it can't be handled when using EVL tail folding. I don't see any problem when using `Data` or `DataWithoutLaneMask`.

**Comment by zixuan-wu** [comment]
> > > I found some loops can't be vectorized and finally I dug to this PR. :-) I simply commented out these lines in this PR and the loop can be vectorized correctly. What's the status now? Are there any things I missed?
> > 
> > 
> > Do you find any other cases with functional/performance issue if it bails out to DataWithoutLaneMask style mask instead of discarding the VPlan? Is it better to bail out to TailFoldingStyle::Data style as what getPreferredTailFoldingStyle prefers when V extension enables. If it bails out to TailFoldingStyle::Data style, then useActiveLaneMask should do some work at tryToBuildVPlanWithVPRecipes.
> 
> The problem I met is because there are some `VPWidenPointerInductionRecipe`s and it can't be handled when using EVL tail folding. I don't see any problem when using `Data` or `DataWithoutLaneMask`.

If bail out to DataWithoutLaneMask, it's still a kind of tail-folding. Is it better than scalar epilogue? As what 'predicate-else-scalar-epilogue' says,  create scalar epilogue if tail folding fails. So the question is whether it's failure tail-folding when DataWithEVL style fails and DataWithoutLaneMask succeed.

**Comment by arcbbb** [comment]
> I found some loops can't be vectorized and finally I dug to this PR. :-) I simply commented out these lines in this PR and the loop can be vectorized correctly. What's the status now? Are there any things I missed?

It depends on the two PRs
- https://github.com/llvm/llvm-project/pull/118638
- https://github.com/llvm/llvm-project/pull/119859

Enabling the vectorization causes issues on HW that implements `ceil(AVL / 2) ≤ vl ≤ VLMAX if AVL < (2 * VLMAX)`
described in https://github.com/riscvarchive/riscv-v-spec/blob/master/v-spec.adoc#sec-vector-config
```
6.3 Constraints on Setting vl
The vset{i}vl{i} instructions first set VLMAX according to their vtype argument, then set vl obeying the following constraints:

vl = AVL if AVL ≤ VLMAX

ceil(AVL / 2) ≤ vl ≤ VLMAX if AVL < (2 * VLMAX)

vl = VLMAX if AVL ≥ (2 * VLMAX)
```

**Comment by wangpc-pp** [comment]
> Enabling the vectorization causes issues on HW that implements `ceil(AVL / 2) ≤ vl ≤ VLMAX if AVL < (2 * VLMAX)`

@arcbbb Actually I don't understand this, can you explain it to me?

**Comment by arcbbb** [comment]
> > Enabling the vectorization causes issues on HW that implements `ceil(AVL / 2) ≤ vl ≤ VLMAX if AVL < (2 * VLMAX)`
> 
> @arcbbb Actually I don't understand this, can you explain it to me?

Under the constraint `ceil(AVL / 2) ≤ VL ≤ VLMAX when AVL < 2 * VLMAX`:
Assuming VLMAX = 16 and AVL = 20, the loop would require 2 iterations.
The values returned by vsetvl in those iterations could be either (16, 4) or (10, 10).
This is HW implementation-defined.

And the increment of widen iv (ptr or integer /fp) is using VLMax instead of VL. it doesn't work on HW implementations which goes (10, 10).
As the diff change shows
```
; CHECK-NEXT:    [[TMP14:%.*]] = call i64 @llvm.vscale.i64()
; CHECK-NEXT:    [[TMP15:%.*]] = mul i64 [[TMP14]], 2
; CHECK-NEXT:    [[TMP16:%.*]] = mul i64 1, [[TMP15]]
; CHECK-NEXT:    [[DOTSPLATINSERT:%.*]] = insertelement <vscale x 2 x i64> poison, i64 [[TMP16]], i64 0
; CHECK-NEXT:    [[DOTSPLAT:%.*]] = shufflevector <vscale x 2 x i64> [[DOTSPLATINSERT]], <vscale x 2 x i64> poison, <vscale x 2 x i32> zeroinitializer
; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
; CHECK:       vector.body:
; .....
; CHECK-NEXT:    [[VEC_IND_NEXT]] = add <vscale x 2 x i64> [[VEC_IND]], [[DOTSPLAT]]
```

**Comment by wangpc-pp** [comment]
@arcbbb Get it! Thanks!

**Comment by zixuan-wu** [comment]
> * [[VPlan] Expand VPWidenIntOrFpInductionRecipe into separate recipes #118638](https://github.com/llvm/llvm-project/pull/118638)
> * [[VPlan] Pass VF as operand to VPWidenPointerInductionRecipe #119859](https://github.com/llvm/llvm-project/pull/119859)
> 

So after those two enhancement, then it will not discard the VPlan?

**Comment by arcbbb** [comment]
> > * [[VPlan] Expand VPWidenIntOrFpInductionRecipe into separate recipes #118638](https://github.com/llvm/llvm-project/pull/118638)
> > * [[VPlan] Pass VF as operand to VPWidenPointerInductionRecipe #119859](https://github.com/llvm/llvm-project/pull/119859)
> 
> So after those two enhancement, then it will not discard the VPlan?

Yes, as long as EVL is applied properly.
